galasa-parent/dev.galasa.framework.api.authentication/src/main/java/dev/galasa/framework/api/authentication/internal/routes/AuthRoute.java

```diff
@@ -0,0 +1,177 @@
+
+                JsonObject responseJson = new JsonObject();
+                responseJson.add("jwt", tokenResponseBodyJson.get(idTokenKey));
+                responseJson.add(refreshTokenKey, tokenResponseBodyJson.get(refreshTokenKey));
+
+                return getResponseBuilder().buildResponse(response, "application/json", gson.toJson(responseJson),
+                        HttpServletResponse.SC_OK);
+            } else {
+                logger.info("Unable to get new bearer and refresh tokens from issuer.");
+            }
+
+        } catch (InterruptedException e) {
+            logger.error("POST request to the OpenID Connect provider's token endpoint was interrupted.", e);
+        }
+
+        ServletError error = new ServletError(GAL5000_GENERIC_API_ERROR);
+        throw new InternalServletException(error, HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+    }
+
+    /**
+     * Gets a given HTTP request's body as a JSON object.
+     */
+    private TokenPayload getRequestBodyAsJson(HttpServletRequest request) throws IOException {
+        StringBuilder sbRequestBody = new StringBuilder();
+        BufferedReader bodyReader = request.getReader();
+
+        String line = bodyReader.readLine();
+        while (line != null) {
+            sbRequestBody.append(line);
+            line = bodyReader.readLine();
+        }
+
+        return gson.fromJson(sbRequestBody.toString(), TokenPayload.class);
+    }
+
+    /**
+     * Sends a POST request to the JWT issuer's /token endpoint and returns the
+     * response's body as a JSON object.
+     *
+     * @param requestBodyJson the request payload containing the required parameters
+     *                        for the /token endpoint
+     */
+    private JsonObject sendTokenPost(HttpServletRequest request, TokenPayload requestBodyJson)
+            throws IOException, InterruptedException {
+        String secret = requestBodyJson.getEncodedSecret();
+        String decodedSecret = new String(Base64.getDecoder().decode(secret));
+
+        // Refresh tokens and authorization codes can be used in exchange for JWTs,
+        // so we need to find out what method was used
+        HttpResponse<String> tokenResponse = null;
+        if (requestBodyJson.getRefreshToken() != null) {
+            tokenResponse = oidcProvider.sendTokenPost(requestBodyJson.getClientId(), decodedSecret, requestBodyJson.getRefreshToken());
+        } else {
+            tokenResponse = oidcProvider.sendTokenPost(requestBodyJson.getClientId(), decodedSecret, requestBodyJson.getCode(), AuthCallbackRoute.getExternalAuthCallbackUrl());
+        }
+        return gson.fromJson(tokenResponse.body(), JsonObject.class);
+    }
+
+    /**
+     * Checks if a given URL is a valid URL.
+     */
+    private boolean isUrlValid(String url) {
+        try {
+            new URL(url).toURI();
+            return true;
+        } catch (URISyntaxException | MalformedURLException e) {
```

Comment:
return at the end of the method using a variable, then we can log the value ?

Reply:
Done